File: tugas linked list kuliah/singly_linked_list.java
import java.util.Scanner;

public class singly_linked_list {

    //static variable (karena function pakai static)
    static int count = 0;
    static Scanner input = new Scanner(System.in);
    static Node head;


    // running program
    public static void main(String[] args) {
    
        Init();
    }
    // class NODE
    static class Node{
        
        int nim;
        String name;
        Node next;

        Node(String name, int nim){
            this.name = name;
            this.nim =  nim;
            this.next =  null;
        }

        // agar ketika print out object di conver ke string , bukan (Node  @12345)
        @Override
        public String toString(){
        return name  + "||" + nim;
    }
    }

    

    // init program
    static void Init(){
        
        int choice = 0;


        System.out.println("choose 1 - 8 for operation and 9 for close program");
        System.out.println("1. insert at beggining");
        System.out.println("2. insert given position");
        System.out.println("3. insert at end");
        System.out.println("4. delete from beggining");
        System.out.println("5. delete given position");
        System.out.println("6. delete frome end");
        System.out.println("7. delete first occurence");
        System.out.println("8. show data");
        System.out.println("9. exit");


        do{
            System.out.print("input only accept 1 - 9: ");

            if(input.hasNextInt()){

            choice = input.nextInt();
            input.nextLine();

            switch(choice){
                case 1:
                    InsertAtBeggining();
                    break;
                case 2:
                    InsertGivenPosition();
                    break;
                case 3:
                    InsertAtEnd();
                    break;
                case 4:
                    DeleteFromBeggining();
                    break;
                case 5:
                    DeleteGivenPosition();
                    break;
                case 6:
                    DeleteFromEnd();
                    break;
                case 7:
                    DeleteFirstOccurence();
                    break;
                case 8:
                    ShowData();
                    break;
                case 9:
                    break;
                default:
                    System.out.println("input only accept 1 - 9!");
            }
        } else {
            System.out.println("can't input string, input again! ");
            input.nextLine();
        }
        
        }while(choice != 9);

        input.close();
    }

    //insert at beggining
    static void InsertAtBeggining (){


            String name;
            int nim;
            
            System.out.print("input name: ");
            name = input.nextLine();

            System.out.print("input nim: ");

            if(input.hasNextInt()){
            nim =  input.nextInt();
            input.nextLine();
            } else {
                System.out.println("nim have to be int");
                input.nextLine();
                return;
            }

            Node newNode = new Node(name, nim);
            newNode.next = head;
            head =  newNode;
            count++;
    }
    

    //show data
    static void ShowData(){
        int i = 1;
        Node temp = head;
        System.out.println(" ");
        System.out.println("count = "+count);
        System.out.print("=========");

        while(temp != null){
            System.out.println( i +". "+ temp.name + " || " + temp.nim);
            temp =  temp.next;
            i++;
        }

        // if(head != null){
        // System.out.println("[][][][][]" + head.name);
        // }
        // System.out.println("head is: "+head);
    }

static void InsertGivenPosition(){
    // Cek apakah linked list kosong
    if(head == null){
        System.out.println("Data is empty");
        return;
    }
    
    // total data di linked list
    Node curr = head;
    int total = 0;
    while(curr != null){
        total++;
        curr = curr.next;
    }
    
    // Tampilkan informasi jumlah data
    System.out.println("Amount data currently: " + total);
    System.out.println("Data from 1- " + (total + 1));
    
    // Input posisi
    int pos = 0;
    System.out.print("Input posisi: ");
    

    // cek input harus int
    if(input.hasNextInt()){
        pos = input.nextInt();
        input.nextLine();
    } else {
        input.nextLine();
        System.out.println("Input only accept number");
        return;
    }
    
    // Validasi posisi
    if(pos < 1 || pos > total + 1){
        System.out.println("Position just accept  1 - " + (total + 1));
        return;
    }
    
    // Input data
    int nim;
    String name;
    System.out.print("Input nama: ");
    name = input.nextLine();
    System.out.print("Input nim: ");
    if(input.hasNextInt()){
    nim = input.nextInt();
    input.nextLine();
    } else {
        System.out.println("input only accept number");
        return;
    }
    // Insert di posisi 1 (awal)
    if(pos == 1){
        Node newNode = new Node(name, nim);
        newNode.next = head;
        head = newNode;
        System.out.println("Data is complete added to 1");
        return;
    }
    
    // Insert di posisi tengah atau akhir
    curr = head;
    for(int i = 1; i < pos - 1 && curr != null; i++){
        curr = curr.next;
    }
    
    Node newNode = new Node(name, nim);
    newNode.next = curr.next;
    curr.next = newNode;
    count++;
    
    System.out.println("Data is complete added at " + pos);
}

    //Insert at end
static void InsertAtEnd(){

    if(head == null){
        System.out.println("head is null");
        return;
    }
    System.out.print("input name: ");
    String name = input.nextLine();
    System.out.print("input nim: ");
    int nim;
    if(input.hasNextInt()){
    nim = input.nextInt();
    input.nextLine();
    } else {
        input.nextLine();
        System.out.println("nim only accept int");
        return;
    
    }

    Node newNode = new Node(name, nim);

    Node last = head;
    while(last.next != null){
        last = last.next;
    }

    last.next = newNode;
    count++;

}
    // Delete from Beggining
    static void DeleteFromBeggining(){

        
        Node curr = head;

        if(curr == null){
            System.out.println("head is null");
            return;
        }
        head = curr.next;
        curr = null;
        count--;
    }


    // Delete From end
    static void DeleteFromEnd(){
        Node curr = head;

        if(curr == null){
            System.out.println("head is null");
            return;
        }

        if(head.next == null){
            head = null;
            count--;
            return;
        }

         while(curr.next.next != null){
            curr = curr.next;
         }
        curr.next = null;
        count--;
    }

    // Delete Given Position
    static void DeleteGivenPosition(){
        
        int pos = 0;

        System.out.print("input pos: ");
        pos = input.nextInt();

        if(pos < 1 || pos > count){
            System.out.println("pos kurang dari 1 dan lebih dari count");
            return;
        }

        if(pos == 1){
            DeleteFromBeggining();
            return;
        }

        if(pos == count){
            DeleteFromEnd();
            return;
        }
        Node curr = head;
        for(int i = 0; i < pos - 2; i++){
            curr = curr.next;
        }
        curr.next = curr.next.next;
        count--;
    }

    // Delete First Occurence
    static void DeleteFirstOccurence() {
    int nim = 0;
    int index = -1;

    
    Node prev = null;
    Node curr = head;

    System.out.print("Input nim: ");


    // finding nim inputed
    if (input.hasNextInt()) {
        nim = input.nextInt();
        input.nextLine();

        for (int i = 0; i < count; i++) {
            if (curr.nim == nim) {
                index = i;
                break;
            }
            prev = curr;       
            curr = curr.next;  
        }

    } else {
        System.out.println("Nim only accepts int");
        input.nextLine();
        return;
    }

    // handle hasil dari index
    if (index == -1) {
        System.out.println("Nim not found");
    } else if (index == 0) {
        DeleteFromBeggining();
    } else if (index == count - 1) {
        DeleteFromEnd();
    } else {
        prev.next = curr.next; 
        count--;
    }
}
}

File: tugas linked list kuliah/singly_linked_list.py
class Node:
    def __init__(self, name: str, nim: int):
        self.name = name
        self.nim = nim
        self.next = None

    def __str__(self):
        return f"{self.name}||{self.nim}"


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.count = 0

    # ─── INSERT ───────────────────────────────────────────────

    def insert_at_beginning(self):
        name = input("Input name: ").strip()

        nim_input = input("Input nim: ").strip()
        if not nim_input.lstrip('-').isdigit():
            print("Nim have to be int")
            return

        nim = int(nim_input)
        new_node = Node(name, nim)
        new_node.next = self.head
        self.head = new_node
        self.count += 1

    def insert_given_position(self):
        if self.head is None:
            print("Data is empty")
            return

        # hitung total node
        curr = self.head
        total = 0
        while curr:
            total += 1
            curr = curr.next

        print(f"Amount data currently: {total}")
        print(f"Data from 1 - {total + 1}")

        pos_input = input("Input posisi: ").strip()
        if not pos_input.lstrip('-').isdigit():
            print("Input only accept number")
            return

        pos = int(pos_input)
        if pos < 1 or pos > total + 1:
            print(f"Position just accept 1 - {total + 1}")
            return

        name = input("Input nama: ").strip()

        nim_input = input("Input nim: ").strip()
        if not nim_input.lstrip('-').isdigit():
            print("Input only accept number")
            return

        nim = int(nim_input)

        if pos == 1:
            new_node = Node(name, nim)
            new_node.next = self.head
            self.head = new_node
            self.count += 1
            print("Data is complete added to 1")
            return

        curr = self.head
        for i in range(1, pos - 1):
            if curr is None:
                break
            curr = curr.next

        new_node = Node(name, nim)
        new_node.next = curr.next
        curr.next = new_node
        self.count += 1
        print(f"Data is complete added at {pos}")

    def insert_at_end(self):
        if self.head is None:
            print("Head is null")
            return

        name = input("Input name: ").strip()

        nim_input = input("Input nim: ").strip()
        if not nim_input.lstrip('-').isdigit():
            print("Nim only accept int")
            return

        nim = int(nim_input)
        new_node = Node(name, nim)

        last = self.head
        while last.next:
            last = last.next

        last.next = new_node
        self.count += 1

    # ─── DELETE ───────────────────────────────────────────────

    def delete_from_beginning(self):
        if self.head is None:
            print("Head is null")
            return

        self.head = self.head.next
        self.count -= 1

    def delete_from_end(self):
        if self.head is None:
            print("Head is null")
            return

        if self.head.next is None:
            self.head = None
            self.count -= 1
            return

        curr = self.head
        while curr.next.next:
            curr = curr.next

        curr.next = None
        self.count -= 1

    def delete_given_position(self):
        pos_input = input("Input pos: ").strip()
        if not pos_input.lstrip('-').isdigit():
            print("Input only accept number")
            return

        pos = int(pos_input)

        if pos < 1 or pos > self.count:
            print("Pos kurang dari 1 dan lebih dari count")
            return

        if pos == 1:
            self.delete_from_beginning()
            return

        if pos == self.count:
            self.delete_from_end()
            return

        curr = self.head
        for i in range(pos - 2):
            curr = curr.next

        curr.next = curr.next.next
        self.count -= 1

    def delete_first_occurrence(self):
        nim_input = input("Input nim: ").strip()
        if not nim_input.lstrip('-').isdigit():
            print("Nim only accepts int")
            return

        nim = int(nim_input)
        index = -1
        prev = None
        curr = self.head

        for i in range(self.count):
            if curr.nim == nim:
                index = i
                break
            prev = curr
            curr = curr.next

        if index == -1:
            print("Nim not found")
        elif index == 0:
            self.delete_from_beginning()
        elif index == self.count - 1:
            self.delete_from_end()
        else:
            prev.next = curr.next
            self.count -= 1

    # ─── SHOW ─────────────────────────────────────────────────

    def show_data(self):
        print()
        print(f"count = {self.count}")
        print("=========")
        temp = self.head
        i = 1
        while temp:
            print(f"{i}. {temp.name} || {temp.nim}")
            temp = temp.next
            i += 1

    # ─── INIT / MENU ──────────────────────────────────────────

    def init(self):
        print("Choose 1 - 8 for operation and 9 to close program")
        print("1. Insert at beginning")
        print("2. Insert given position")
        print("3. Insert at end")
        print("4. Delete from beginning")
        print("5. Delete given position")
        print("6. Delete from end")
        print("7. Delete first occurrence")
        print("8. Show data")
        print("9. Exit")

        choice = 0
        while choice != 9:
            raw = input("Input only accept 1 - 9: ").strip()

            if not raw.lstrip('-').isdigit():
                print("Can't input string, input again!")
                continue

            choice = int(raw)

            if choice == 1:
                self.insert_at_beginning()
            elif choice == 2:
                self.insert_given_position()
            elif choice == 3:
                self.insert_at_end()
            elif choice == 4:
                self.delete_from_beginning()
            elif choice == 5:
                self.delete_given_position()
            elif choice == 6:
                self.delete_from_end()
            elif choice == 7:
                self.delete_first_occurrence()
            elif choice == 8:
                self.show_data()
            elif choice == 9:
                break
            else:
                print("Input only accept 1 - 9!")


if __name__ == "__main__":
    sll = SinglyLinkedList()
    sll.init()